Review followup: PR #21421 — t2997: fix BSD-incompatible mktemp templates (XXXXXX before .ext)

## Unaddressed review bot suggestions

PR #21421 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/21421

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `prompts/build.txt`
section 6a). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `prompts/build.txt` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/lint-mktemp-portability.sh:145`
- `.agents/scripts/lint-mktemp-portability.sh:174`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/lint-mktemp-portability.sh:145`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21421#discussion_r3150447955)

```diff
+
+# ---------------------------------------------------------------------------
+# File list resolution
+# ---------------------------------------------------------------------------
+if [[ ${#TARGET_FILES[@]} -eq 0 ]]; then
+	# Default: all tracked .sh files in the repo.
+	if ! command -v git >/dev/null 2>&1; then
+		printf 'mktemp-portability: git not found and no FILES given\n' >&2
+		exit 2
+	fi
+	# shellcheck disable=SC2207
+	TARGET_FILES=($(git ls-files '*.sh' 2>/dev/null || true))
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> There are two problems with this file discovery logic:
> 1. **Word Splitting**: Using `$(...)` inside an array assignment without `mapfile` or `readarray` will cause filenames containing spaces to be split into multiple array elements.
> 2. **Incomplete Search**: The `git ls-files` command only looks for `*.sh` files, but the `scan_file` function (lines 181-184) is explicitly designed to handle `.bash` and `.zsh` files as well. These will be ignored by default.
> 
> Consider using a `while read` loop to safely populate the array with all relevant shell script extensions.
> 
> ```suggestion
> while IFS= read -r f; do TARGET_FILES+=("$f"); done < <(git ls-files '*.sh' '*.bash' '*.zsh' 2>/dev/null || true)
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Avoid using the word 'issue' as it is too generic and can lead to false positives by matching phrases like 'no issues found'.
> </details>

#### gemini-code-assist on `.agents/scripts/lint-mktemp-portability.sh:174`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21421#discussion_r3150447963)

```diff
+# We use POSIX-friendly grep with -P (PCRE) since rg may not be on the runner.
+# Fall back to grep -P if rg unavailable.
+#
+# Suppression: a line with the comment `# mktemp-portability: ignore` is excluded.
+
+violations_found=0
+violation_lines=()
+
+# Build the search pattern. Use a literal-X form to avoid pattern-detection
+# hitting the lint script itself.
+_pattern='mktemp[[:space:]]+(?!-[duqQ][[:space:]])[^|;&)`]*'
+_pattern+='X{6}\.[a-zA-Z]'
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The negative lookahead `(?!-[duqQ][[:space:]])` is fragile. It only matches if the flag is the first argument and is not combined with other flags (e.g., it fails to skip `mktemp -dq ...` or `mktemp -qu ...`). It also doesn't account for multiple spaces between the command and the flag.
> 
> A more robust pattern would check for the presence of these flags anywhere in the argument string before the template placeholder.
> 
> ```suggestion
> _pattern='mktemp[[:space:]]+(?![^|;&)]*[[:space:]]-[a-zA-Z]*[duqQ][a-zA-Z]*\b)[^|;&)]*'
> _pattern+='X{6,}\.[a-zA-Z]'
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. In shell scripts, use POSIX character classes like [[:space:]] instead of non-standard shortcuts like \s for better portability across regex engines.
> </details>

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/21421#pullrequestreview-4184477536)

> ## Code Review
> 
> This pull request addresses macOS portability issues with `mktemp` by ensuring that template arguments end with the `XXXXXX` placeholder, as BSD `mktemp` fails to substitute placeholders followed by extensions. The changes include updates to the shell style guide, refactoring of multiple scripts to use compliant `mktemp` patterns, and the introduction of a new linting script, `lint-mktemp-portability.sh`. Feedback on the linting script identifies issues with file discovery logic regarding word splitting and missing file extensions, as well as fragility in the regex used to detect and skip specific `mktemp` flags.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.13.5 automated scan.

--- a/.agents/scripts/lint-mktemp-portability.sh
+++ b/.agents/scripts/lint-mktemp-portability.sh
@@ -32,7 +32,7 @@
 #   --no-exit-code  always exit 0 (advisory mode)
 #   --help          show this help
 #
-# With no FILES: scans all git-tracked .sh files via `git ls-files '*.sh'`.
+# With no FILES: scans all git-tracked .sh, .bash, and .zsh files.
 #
 # Output format: file:line: <offending mktemp template>
 # Exit codes: 0=clean, 1=violations found
@@ -141,8 +141,8 @@ if [[ ${#TARGET_FILES[@]} -eq 0 ]]; then
 		printf 'mktemp-portability: git not found and no FILES given\n' >&2
 		exit 2
 	fi
-	# shellcheck disable=SC2207
-	TARGET_FILES=($(git ls-files '*.sh' 2>/dev/null || true))
+	while IFS= read -r _f; do TARGET_FILES+=("$_f"); done \
+		< <(git ls-files '*.sh' '*.bash' '*.zsh' 2>/dev/null || true)
 fi
 
 if [[ ${#TARGET_FILES[@]} -eq 0 ]]; then
@@ -155,7 +155,8 @@ fi
 # ---------------------------------------------------------------------------
 # The detection regex:
 #   - mktemp                 — the command
-#   - (?!\s+-[duqQ]\b)       — NOT followed by -d / -u / -q / -Q (file-form only)
+#   - (?!-[a-zA-Z]*[duqQ][a-zA-Z]*[[:space:]]) — NOT if first flag contains
+#                              d/u/q/Q (handles combined flags like -dq, -qu).
 #                              NOTE: -t (template prefix) is fine — its template
 #                              still goes through the same X-placement rule.
 #   - .*X{6,}\.[a-zA-Z]      — XXXXXX (or more) followed by `.<letter>` extension
@@ -170,8 +171,8 @@ violation_lines=()
 
 # Build the search pattern. Use a literal-X form to avoid pattern-detection
 # hitting the lint script itself.
-_pattern='mktemp[[:space:]]+(?!-[duqQ][[:space:]])[^|;&)`]*'
-_pattern+='X{6}\.[a-zA-Z]'
+_pattern='mktemp[[:space:]]+(?!-[a-zA-Z]*[duqQ][a-zA-Z]*[[:space:]])[^|;&)`]*'
+_pattern+='X{6,}\.[a-zA-Z]'
 
 scan_file() {
 	local file="$1"

